### PR TITLE
Fix defaultTags to trigger downstream updates

### LINKED ISSUE #1655: unexpected behavior of defaultTags
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

The AWS provider ignores changes to defaultTags after a resource has been created unless other attributes are being changed

### Steps to reproduce

1. Specify some defaultTags and create a bucket. The bucket has the tags as expected:

```
~/tmp/wtf/defaultTags$ cat index.ts
import * as aws from "@pulumi/aws";

const bucket = new aws.s3.Bucket("bucket-with-default-tags", {bucket: "bucket-with-default-tags"});
~/tmp/wtf/defaultTags$ cat Pulumi.dev.yaml
encryptionsalt: v1:ggyom6IhdG8=:v1:sJqljlbglc8JJgpe:jYzdmym/+StNPwPbn+7uIpAIL+fKLA==
config:
  aws:region: eu-west-1
  aws:defaultTags:
    tags:
      Tag1: value1
      Tag2: value2
~/tmp/wtf/defaultTags$ pulumi up --diff
Previewing update (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
    + aws:s3/bucket:Bucket: (create)
        [urn=urn:pulumi:dev::defaultTags::aws:s3/bucket:Bucket::bucket-with-default-tags]
        [provider=urn:pulumi:dev::defaultTags::pulumi:providers:aws::default_4_23_0::04da6b54-80e4-46f7-96ec-b56ff0331ba9]
        acl         : "private"
        bucket      : "bucket-with-default-tags"
        forceDestroy: false

import * as aws from "@pulumi/aws";
Resources:
    + 1 to create
    1 unchanged
Do you want to perform this update? yes
Updating (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
    + aws:s3/bucket:Bucket: (create)
        [urn=urn:pulumi:dev::defaultTags::aws:s3/bucket:Bucket::bucket-with-default-tags]
        [provider=urn:pulumi:dev::defaultTags::pulumi:providers:aws::default_4_23_0::8510e26a-009d-460f-a5e1-4f5e95a6af22]
        acl         : "private"
        bucket      : "bucket-with-default-tags"
        forceDestroy: false
Resources:
    + 1 created
    1 unchanged

Duration: 11s
~/tmp/wtf/defaultTags$ aws s3api get-bucket-tagging --bucket bucket-with-default-tags
{
    "TagSet": [
        {
            "Key": "Tag1",
            "Value": "value1"
        },
        {
            "Key": "Tag2",
            "Value": "value2"
        }
    ]
}
```

2. Add some new tags to the defaultTags and attempt an update, which shows no diff and results in no changes:

```
~/tmp/wtf/defaultTags$ echo '      Tag3: value3' >> Pulumi.dev.yaml
~/tmp/wtf/defaultTags$ echo '      Tag4: value4' >> Pulumi.dev.yaml
~/tmp/wtf/defaultTags$ cat Pulumi.dev.yaml
encryptionsalt: v1:ggyom6IhdG8=:v1:sJqljlbglc8JJgpe:jYzdmym/+StNPwPbn+7uIpAIL+fKLA==
config:
  aws:region: eu-west-1
  aws:defaultTags:
    tags:
      Tag1: value1
      Tag2: value2
      Tag3: value3
      Tag4: value4
~/tmp/wtf/defaultTags$ pulumi up --diff
Previewing update (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
Resources:
    2 unchanged
Do you want to perform this update? yes
Updating (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
Resources:
    2 unchanged

Duration: 3s
~/tmp/wtf/defaultTags$ aws s3api get-bucket-tagging --bucket bucket-with-default-tags
{
    "TagSet": [
        {
            "Key": "Tag1",
            "Value": "value1"
        },
        {
            "Key": "Tag2",
            "Value": "value2"
        }
    ]
}
```

3. Change another property of the bucket and observe how the new tags are also added to the bucket (but are not being shown in the preview):

```
~/tmp/wtf/defaultTags$ vim index.ts
~/tmp/wtf/defaultTags$ cat index.ts
import * as aws from "@pulumi/aws";

const bucket = new aws.s3.Bucket("bucket-with-default-tags", {
    bucket: "bucket-with-default-tags",
    acl: "authenticated-read",
});
~/tmp/wtf/defaultTags$ pulumi up --diff
Previewing update (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
    ~ aws:s3/bucket:Bucket: (update)
        [id=bucket-with-default-tags]
        [urn=urn:pulumi:dev::defaultTags::aws:s3/bucket:Bucket::bucket-with-default-tags]
        [provider=urn:pulumi:dev::defaultTags::pulumi:providers:aws::default_4_23_0::8510e26a-009d-460f-a5e1-4f5e95a6af22]
      ~ acl: "private" => "authenticated-read"
Resources:
    ~ 1 to update
    1 unchanged
Do you want to perform this update? yes
Updating (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
    ~ aws:s3/bucket:Bucket: (update)
        [id=bucket-with-default-tags]
        [urn=urn:pulumi:dev::defaultTags::aws:s3/bucket:Bucket::bucket-with-default-tags]
        [provider=urn:pulumi:dev::defaultTags::pulumi:providers:aws::default_4_23_0::8510e26a-009d-460f-a5e1-4f5e95a6af22]
      ~ acl: "private" => "authenticated-read"
        --outputs:--
      ~ tagsAll                          : {
            Tag1: "value1"
            Tag2: "value2"
          + Tag3: "value3"
          + Tag4: "value4"
        }
Resources:
    ~ 1 updated
    1 unchanged

Duration: 15s
~/tmp/wtf/defaultTags$ aws s3api get-bucket-tagging --bucket bucket-with-default-tags
{
    "TagSet": [
        {
            "Key": "Tag1",
            "Value": "value1"
        },
        {
            "Key": "Tag4",
            "Value": "value4"
        },
        {
            "Key": "Tag3",
            "Value": "value3"
        },
        {
            "Key": "Tag2",
            "Value": "value2"
        }
    ]
}
~/tmp/wtf/defaultTags$
```

Expected:

- I expected the Tag3 and Tag4 to be added to the bucket in step 2.
- I also expected the preview in step 3 to show the addition of Tag3 and Tag4 given that it has added them in the update.

Actual: 

- Tag3 and Tag4 were not added in step 2
- Addition of Tag3 and Tag4 was not shown in preview of step 3

Inspiration for this issue report: https://stackoverflow.com/a/68829181/13846505

### PR DESCRIPTION
**Pull request by AaronFriel**
The `defaultTags` input property on providers now triggers an update on downstream resources. Because of the nature of the fix in the Pulumi TF Bridge, and to limit the scope of the change, we mark only the `tags_all` field from TF schema as a `ComputedInput`, triggering new diff behavior within the bridge.

As the `tags_all` field is otherwise hidden as a "computed" value, we omit detailed diff information. This is a tradeoff to avoid a more invasive change with more risk.

Fixes #1655. To use a release version of TF Bridge, depends on:
- https://github.com/pulumi/pulumi-terraform-bridge/pull/1251

Example output:

```
pulumi up --skip-preview
Updating (test-aws-1655)

View in Browser (Ctrl+O): https://app.pulumi.com/pulumi/simple-yaml/test-aws-1655/updates/9

     Type                     Name                       Status              Info
     pulumi:pulumi:Stack      simple-yaml-test-aws-1655
 ~   ├─ pulumi:providers:aws  aws-provider               updated (0.13s)     [diff: ~defaultTags]
 ~   ├─ aws:s3:Bucket         my-bucket                  updated (0.87s)
 ~   └─ aws:s3:BucketObject   index.html                 updated (0.24s)
```

Verified when adding, updating, deleting, and removing entirely the `defaultTags` property.

### CODE REVIEW

File: provider/resources.go [hunk 1]
```diff
@@ -7004,5 +7004,21 @@ func Provider() tfbridge.ProviderInfo {
...
+			if prov.Resources[key].Fields == nil {
+				prov.Resources[key].Fields = make(map[string]*tfbridge.SchemaInfo)
+			}
+			if f := prov.Resources[key].Fields["tags_all"]; f != nil {
+				f.ComputedInput = true
+			} else {
+				prov.Resources[key].Fields["tags_all"] = &tfbridge.SchemaInfo{
+					ComputedInput: true,
+				}
+			}
```
Comment (iwahbe) [inline provider/resources.go:7018]: This is just more consistent.
```suggestion
			if f := prov.Resources[key].Fields["tags_all"]; f == nil {
				prov.Resources[key].Fields["tags_all"] = make(tfbridge.SchemaInfo)
			}
			prov.Resources[key].Fields["tags_all"].ComputedInput=  true
```

File: provider/resources.go [hunk 2]
```diff
@@ -7004,5 +7004,21 @@ func Provider() tfbridge.ProviderInfo {
 	// Add a CSharp-specific override for aws_s3_bucket.bucket.
 	prov.Resources["aws_s3_bucket_legacy"].Fields["bucket"].CSharpName = "BucketName"
 
+	prov.P.ResourcesMap().Range(func(key string, value shim.Resource) bool {
```
Comment (iwahbe) [inline provider/resources.go:7007]: Do we expect this to apply to all fields called `"tags_all"`? If so, I would love a comment to explain that `"tags_all" is commonly used to inherit tags, and so ...`. If not, can we narrow this to only apply to the closed set of resources where this is a problem.

**Review by iwahbe — APPROVED** [review]

File: provider/resources.go (hunk 2 above)
Comment (AaronFriel) [reply provider/resources.go:7007]: Updated this to match the upstream provider, which checks for a computed tags_all and a non-computed tags: https://github.com/pulumi/pulumi-aws/blob/a82c36c5556e2b8de91ba2a82a24bcd6074345be/internal/provider/provider.go#L394-L415

**Review by t0yv0 — CHANGES_REQUESTED** [review]
I have rebuild the provider on top of a custom terraform-bridge with this line removed:

```
 		case v.IsArray():
 			// If this value has a diff and is considered computed by Terraform, the diff will be woefully incomplete. In
 			// this case, do not recurse into the array; instead, just use the count diff for the details.
-			if d := tfDiff.Attribute(name + ".#"); d == nil || (!d.NewComputed && !isComputedInput) {
+			if d := tfDiff.Attribute(name + ".#"); d == nil || !d.NewComputed {

```

And the new integration test still passes.

Could we please consider removing the mystery change? Or is there an explanation, if not a repro, why is it needed?

**Review by t0yv0 — APPROVED** [review]
LGTM! But it should have a review from the providers team before going in.

**Review by kpitzen — APPROVED** [review]

File: sdk/nodejs/provider.ts [hunk 3]
```diff
@@ -27,7 +27,7 @@ export class Provider extends pulumi.ProviderResource {
         if (obj === undefined || obj === null) {
             return false;
         }
-        return obj['__pulumiType'] === Provider.__pulumiType;
+        return obj['__pulumiType'] === "pulumi:providers:" + Provider.__pulumiType;
```
Comment (kpitzen) [inline sdk/nodejs/provider.ts:30]: 👀

**Comment by AaronFriel** [comment]
All checks have passed, confirmed via checks tab in GitHub UI. Using admin to merge

**Pull request merged**